### PR TITLE
Update BRE android

### PR DESCRIPTION
**Pull request by deeppandya**
Resolves https://github.com/brave/brave-browser/issues/12488 https://github.com/brave/brave-browser/issues/13008 https://github.com/brave/brave-browser/issues/12992

## Submitter Checklist:

- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed).
- [ ] Requested a security/privacy review as needed.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

- Follow https://www.figma.com/file/0dbMYs8Wp8W3RgrG1yTdki/RE-1.0?node-id=541%3A19529 for rewards onboarding
- Verify the **Skip** button functionality
- For smaller screen, rewards modal and onboarding should have vertical scrolling enabled and scroller should be visible.

### CODE REVIEW

File: android/java/org/chromium/chrome/browser/app/BraveActivity.java [hunk 1]
```diff
@@ -97,8 +99,12 @@
     public static final int USER_WALLET_ACTIVITY_REQUEST_CODE = 35;
     public static final String ADD_FUNDS_URL = "chrome://rewards/#add-funds";
     public static final String REWARDS_SETTINGS_URL = "chrome://rewards/";
+    public static final String BRAVE_REWARDS_SETTINGS_URL = "brave://rewards/";
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/app/BraveActivity.java:102]: do we need that one and one above duplicated?

File: android/java/org/chromium/chrome/browser/onboarding/OnboardingPrefManager.java [hunk 2]
```diff
@@ -194,7 +194,6 @@ public void setOnboardingShownForSkip(boolean isShown) {
     public boolean showOnboardingForSkip(Context context) {
         boolean shouldShow = PackageUtils.isFirstInstall(context)
                              && !hasOnboardingShownForSkip()
-                             && (ChromeFeatureList.isEnabled(BraveFeatureList.BRAVE_REWARDS) && !UserPrefs.get(Profile.getLastUsedRegularProfile()).getBoolean(BravePref.ENABLED))
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/onboarding/OnboardingPrefManager.java:197]: are we fine with x86 devices? as this check `ChromeFeatureList.isEnabled(BraveFeatureList.BRAVE_REWARDS)` was to exclude them as no rewards there.

File: android/java/org/chromium/chrome/browser/toolbar/top/BraveToolbarLayout.java [hunk 3]
```diff
@@ -465,7 +594,6 @@ public void onClick(View v) {
 
   private boolean checkForRewardsOnboarding() {
     return PackageUtils.isFirstInstall(getContext())
-           && (ChromeFeatureList.isEnabled(BraveFeatureList.BRAVE_REWARDS) && !UserPrefs.get(Profile.getLastUsedRegularProfile()).getBoolean(BravePref.ENABLED))
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/toolbar/top/BraveToolbarLayout.java:468]: the same question about x86

File: android/java/org/chromium/chrome/browser/app/BraveActivity.java (hunk 1 above)
Comment (deeppandya) [reply android/java/org/chromium/chrome/browser/app/BraveActivity.java:102]: @SergeyZhukovsky to show a custom dialog in rewards settings page, we need to check for both urls as you can go to the same page by same url. so i had to add it here

File: android/java/org/chromium/chrome/browser/onboarding/OnboardingPrefManager.java (hunk 2 above)
Comment (deeppandya) [reply android/java/org/chromium/chrome/browser/onboarding/OnboardingPrefManager.java:197]: good catch. i forgot about x86 devices. i will revert it.

File: android/java/org/chromium/chrome/browser/app/BraveActivity.java (hunk 1 above)
Comment (SergeyZhukovsky) [reply android/java/org/chromium/chrome/browser/app/BraveActivity.java:102]: sounds good

**Review by SergeyZhukovsky — APPROVED** [review]
++

File: android/java/org/chromium/chrome/browser/BraveRewardsHelper.java [hunk 4]
```diff
@@ -42,8 +44,17 @@
 import java.util.Locale;
 
 public class BraveRewardsHelper implements LargeIconBridge.LargeIconCallback{
-    private static final String PREF_BRAVE_REWARDS_OPEN_COUNT = "brave_rewards_open_count";
-    private static final String PREF_SHOW_BRAVE_REWARDS_ONBOARDING = "show_brave_rewards_onboarding";
+    private static final String PREF_BRAVE_REWARDS_OPEN_COUNT = "brave_rewards_app_open_count";
```
Comment (emerick) [inline android/java/org/chromium/chrome/browser/BraveRewardsHelper.java:47]: Maybe update the identifier to reflect the new preference name too?

File: android/java/org/chromium/chrome/browser/ntp_background_images/util/NTPUtil.java [hunk 5]
```diff
@@ -170,11 +170,16 @@ public static void showNonDistruptiveBanner(ChromeActivity chromeActivity, View
...
             @Override
             public void onClick(View view) {
-                // clickOnBottomBanner(chromeActivity, ntpType, nonDistruptiveBannerLayout, sponsoredTab, newTabPageListener);
-                if (BraveActivity.getBraveActivity() != null) {
-                    nonDistruptiveBannerLayout.setVisibility(View.GONE);
-                    BraveRewardsHelper.setShowBraveRewardsOnboarding(true);
-                    BraveActivity.getBraveActivity().openRewardsPanel();
+                if (BraveAdsNativeHelper.nativeIsBraveAdsEnabled(
+                            Profile.getLastUsedRegularProfile())) {
+                    clickOnBottomBanner(chromeActivity, ntpType, nonDistruptiveBannerLayout,
```
Comment (emerick) [inline android/java/org/chromium/chrome/browser/ntp_background_images/util/NTPUtil.java:175]: Nit: Seems like `nonDistruptiveBannerLayout` has a typo (extra 't' in `Disruptive`)?

**Review by emerick — APPROVED** [review]
LGTM, just found a couple of nits...

File: browser/brave_rewards/android/brave_rewards_native_worker.cc [hunk 6]
```diff
@@ -273,6 +273,24 @@ void BraveRewardsNativeWorker::FetchGrants(JNIEnv* env,
   }
 }
 
+void BraveRewardsNativeWorker::StartProcess(
+    JNIEnv* env,
+    const base::android::JavaParamRef<jobject>& obj) {
+  if (brave_rewards_service_) {
+    brave_rewards_service_->StartProcess(
+        base::BindOnce(&BraveRewardsNativeWorker::OnStartProcess,
+                       base::Unretained(this), brave_rewards_service_));
```
Comment (zenparsing) [inline browser/brave_rewards/android/brave_rewards_native_worker.cc:282]: Do we need to bind `brave_rewards_service_` here? It doesn't look like it's used in the callback (and the callback already has access to `brave_rewards_service_` via `this`).

**Comment by LaurenWags** [comment]
Verified on LG Nexus 5 with Android 5.1 running 1.18.67 Bravearm.apk

Verified test plan from https://github.com/brave/brave-core/pull/7378#issue-534456802

Generally confirmed `1.18 Android` section of the linked figma document - https://www.figma.com/file/0dbMYs8Wp8W3RgrG1yTdki/RE-1.0?node-id=541%3A19529. Note, generally confirmed onboarding looks good. More detailed testing will be done when this is uplifted to 1.18.x. Confirmed the following items:

<details>
<summary>Entry Points - New User - Panel onboarding</summary>

Confirmed tapping on the BAT logo showed the large onboarding modal expected.
Screenshot_2020-12-10-10-17-50

</details>

<details>
<summary>Entry Points - New User - Rewards Settings onboarding</summary>

Confirmed finding BR via 3 dots menu > Brave Rewards and opening the page showed the large onboarding modal as expected. 
Screenshot_2020-12-10-10-20-24

</details>

<details>
<summary>Onboarding Sequence - New User - First App Open</summary>

Confirmed if initial onboarding is "x"-ed out, the smaller UI is surfaced. Confirmed "x", "Quick Tour", "Terms of Service", "Privacy Policy", and "I'm in!" functioned as expected on smaller UI. Note, tapping on "I'm in!" Also opens chrome://rewards in the tab behind the panel. Per discussion this is expected.
Screenshot_2020-12-10-10-24-54

Confirmed if I tap on "Take a quick tour" I get the expected onboarding flow. Confirmed once the onboarding tour is complete the small UI is surfaced.
Screenshot_2020-12-10-14-57-05

Confirmed if I tap on "Start using Brave Rewards" I get the expected onboarding flow. Note - since it takes a minute or so to get ads enabled, you may not get the Settings slide in this onboarding flow. If that happens, try again and go a bit slower through onboarding. Also, there is an issue with AC amount not being set correctly, logged as https://github.com/brave/brave-browser/issues/13148. Confirmed expected panel UI is displayed after this onboarding tour is complete.
Screenshot_2020-12-10-15-09-09

</details>

<details>
<summary>Onboarding Sequence - New User - on next app and panel opening</summary>

Confirmed if I don't opt in to BR, when I launch the app a second time and open the panel I get the smaller opt-in UI.
Screenshot_2020-12-10-15-22-41

</details>

<details>
<summary>Onboarding Sequence - New User - opt-ed in to BR in previous browser session</summary>

Confirmed if I do opt in to BR, when I launch the app a second time and open the panel I get the "Welcome to Brave Rewards" UI w/ option for refresher tour + ads payout info.
Screenshot_2020-12-10-15-24-38

</details>

<details>
<summary>Day 4 - Forced Rewards onboarding</summary>

1. Confirmed no interaction with Panel within first 24h, does not automatically open panel for onboarding. Closed app and removed from memory.
2. Advanced date/time by 24h. Confirmed panel did not automatically open. Closed app and removed from memory.
3. Advanced date/time by 24h. Confirmed panel did not automatically open. Closed app and removed from memory.
4. Advanced date/time by 24h. Confirmed panel did not automatically open. Closed app and removed from memory.
5. Advanced date/time by 24h. On launch panel opened and showed large onboarding modal as expected. Confirmed able to interact with modal.

Screenshot_2020-12-10-10-17-50

</details>

Cc @kjozwiak for the following onboarding flow checks:
Entry Points - New User  - NTP SI
Entry Points - Existing User, Ads on - NTP SI
Onboarding Sequence - New User - First SI Image

Also confirmed the following:

<details>
<summary>Onboarding Skip button works as expected</summary>

Confirmed selecting "Skip" on the first slide/card of onboarding skips all the middle slides/cards and instead directs the user to the final one.
Screenshot_2020-12-10-15-01-24

</details>

<details>
<summary>Scroll bar is shown on small screens</summary>

Confirmed on a smaller screen the initial slide/card of onboarding shows a scroll bar so users know that there is more content (the Skip link) below the visible area.
Screenshot_2020-12-10-16-11-15

</details>

<details>
<summary>Claiming promotions</summary>

Confirmed able to get UGP grant (staging) without having to enable Rewards, however I crash when attempting to claim the grant. Logged https://github.com/brave/brave-browser/issues/13150.

</details>

Confirmed https://github.com/brave/brave-browser/issues/12992 is resolved. Confirmed STR in issue no longer reproduce the problem. Onboarding is displayed appropriately.

Confirmed https://github.com/brave/brave-browser/issues/13008 is resolved. chrome://rewards page displays onboarding as per above scenario.

Confirmed https://github.com/brave/brave-browser/issues/12488 is resolved. Publisher info displays in the panel without having to join Rewards or visit chrome://rewards page. Note - it takes a few seconds on the first site, but the publisher info does display. Subsequent sites visited do not experience a delay.


Issues:
AC amount selection on settings slide does not work --> always selects 10 BAT no matter what I select on the slide. Logged https://github.com/brave/brave-browser/issues/13148.
Crash when claiming promotions. Logged https://github.com/brave/brave-browser/issues/13150.

**Comment by kjozwiak** [comment]
Using https://www.figma.com/file/0dbMYs8Wp8W3RgrG1yTdki/RE-1.0?node-id=541%3A19529, checked the following flows:
* Entry Points - New User - NTP SI
* Entry Points - Existing User, Ads on - NTP SI
* Onboarding Sequence - New User - First SI Image
* Day 4 - Forced Rewards Onboarding

<details>
<summary>Entry Points - New User - NTP SI</summary>

* ensured that dismissing the the NTP SI rewards modal via `X` doesn't launch/start rewards panel/onboarding
* ensured that tapping on the rewards modal when first viewing a NTP SI opens the rewards panel
* ensured that the rewards panel is opened via NTP SI even though it was dismissed via other entry points. Example:
  * tap on the rewards icon via the URL and dismiss onboarding
  * tap on the rewards NTP SI modal and ensure that rewards panel is opened with the new onboarding

<p align="center">

</p>
</details>

<details>
<summary> Entry Points - Existing User, Ads on - NTP SI</summary>

* ensured that dismissing the the NTP SI rewards modal via `X` closes the modal without any issues
* ensured that the NTP SI modal expands/displays the correct information if ads have already been enabled
* ensured that tapping on `Learn more about sponsored images` opens a new tab with `brave.com/faq-rewards`
* ensured that tapping on `Hide sponsored images` disabled NTP SI and they stop appearing when opening NTP

<p align="center">

</p>
</details>

<details>
<summary> Onboarding Sequence - New User - First SI Image</summary>

* ensured that tapping on `Terms of Services` & `Privacy Policy` via the rewards panel doesn't dismiss onboarding
* ensured that tapping on `Start using Brave Rewards` starts the onboarding process
  * ensure that you can navigate through the onboarding experience using either of the following:
    * `< GO BACK`, `NEXT >` or swiping side to side
  * ensured tapping on `Done` completes the onboarding flow
  * ensured that `Skip` takes you to the last page/panel (Wooohoooo) without any issues
  * ensured that tapping on `X` closes onboarding
* ensured that tapping on `Take a quick tour` starts the rewards tour (same as onboarding without ads/a-c opt-in)
  * ensure that you can navigate through the onboarding experience using either of the following:
    * `< GO BACK`, `NEXT >` or swiping side to side
  * ensured tapping on `Done` completes the tour
  * ensured that `Skip` takes you to the last page/panel (Wooohoooo) without any issues
  * ensured that tapping on `X` closes onboarding

**`Example`** | **`Example`**
---------------|----------------
Screenshot_20201211-012654_Brave - Nightly | Screenshot_20201211-011703_Brave - Nightly
</details>

<details>
<summary>Day 4 - Forced Rewards Onboarding</summary>

* ensured the the rewards panel automatically opens after 4 days if the user hasn't opened the rewards panel or NTP SI panel
* ensured that the rewards panel doesn't open on the fourth day if the user has already interacted with it via the NTP SI

<p align="center">

</p>
</details>

Issues found while going through the above:

* reward payouts information via onboarding differs from rewards panel - https://github.com/brave/brave-browser/issues/13158
* rewards panel being opened after 4 days even though it was viewed/dismissed via NTP SI entry point #13160 - https://github.com/brave/brave-browser/issues/13160

CCing @deeppandya @codybanxs @LaurenWags

**Pull request merged**